fix: allow empty list of NICs for Underlay setup if Multus is used

fdomain opened the pull request:
* specifying a NIC in the Underlay configuration was incorrectly mandatory, which prevented users from leveraging Multus to establish connectivity to the ToR.
* we now detect if the Router Pod has a `k8s.v1.cni.cncf.io/networks` annotation and propagate that information to subsequent calls.
* if no underlay NIC is specified and the multusEnabled flag is true, the logic for retrieving or moving the underlay interface into the router's netns is skipped, assuming that Multus has already handled the necessary setup.

<!-- Thanks for sending a pull request!
1. If this is your first time, please read the [contributing guide](https://https://openperouter.github.io/docs/contributing/)
2. For non-trivial pull requests, please [file an
   issue](https://github.com/openperouter/openperouter/issues/new) first, and get
   agreement that the change is a good idea, and a general guideline
   for how it should be implemented, before sending code. Large PRs
   that weren't first discussed and agreed upon in an issue won't be
   accepted.
3. If the PR fixes a particular bug, please include the words "Fixed
   #<issue number>" in the PR text, so that the bug auto-closes when
   the PR is merged.
-->

**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind cleanup
> /kind feature
> /kind design
> /kind flake
> /kind failing
> /kind documentation
> /kind regression
> /kind example

**What this PR does / why we need it**:
Fixes #150 

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If no release note is required, just write "NONE".
-->

```release-note
Fix: allow omitting underlay NIC configuration when using Multus.
```

fdomain commented:
@fedepaol the e2etests are passing locally on my laptop (46/46). Could it be some flakiness in the tests? Would be great if you could pull this PR on your side and test🙏

fedepaol commented:
> @fedepaol the e2etests are passing locally on my laptop (46/46). Could it be some flakiness in the tests? Would be great if you could pull this PR on your side and test🙏

CI had some flakes _but not this much_. I'll test this on my side, in the meantime you can file a dummy PR to trigger CI and see where we are. Also, the status of the cluster (including CRs, pod logs and so on) after a test failure can be found in the gh action https://github.com/openperouter/openperouter/actions/runs/18949481172 , maybe it's worth a look.
Our e2es dump the state after each failure, so it should be easy (depending on the failure) to find the problem.

fedepaol commented:
For the record, both this https://github.com/openperouter/openperouter/pull/161 and https://github.com/openperouter/openperouter/pull/160 are fully green (not touching anything)

fdomain commented:
I found the issue and fixed it. The problem was that router pods were no longer restarting when underlay objects were deleted. This happened because I introduced a bug: the restart logic depends on an UnderlayRemovedError{} being thrown, but I mistakenly added a condition around the code handling that case. During e2e tests, the router pods were never restarted.
See: https://github.com/openperouter/openperouter/compare/a2e0e8c96015376b1f3079904715003d4cde0cd9..18f0ba17b18538831da5c087aa2fa9252725d5e5

So now the original behaviour is restored:
- if a Nic was specified in an Underlay object, and the Underlay is deleted => `hasAlreadyUnderlay` will be true and the error will be raised since `len(underlays) == 0`
- if there was no NIC specified, and the Underlay access was managed by a Multus annotation => `hasAlreadyUnderlay` will be false, and the pod won't be restarted since the configuration of that interface is not managed by the Underlay object.

fedepaol commented on internal/controller/routerconfiguration/underlay_vni_controller.go:
```diff
@@ -81,6 +81,10 @@ func (r *PERouterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, err
 	}
+	multusEnabled := false
+	if _, ok := routerPod.Annotations["k8s.v1.cni.cncf.io/networks"]; ok {
```
what if we make this a parameter? We'll have to drive the annotation from the chart anyways.

fedepaol commented on internal/hostnetwork/underlay_test.go:
```diff
@@ -149,19 +149,32 @@ var _ = Describe("Underlay configuration should work when", func() {
 })
 
-func validateUnderlayInNS(g Gomega, ns netns.NsHandle, params UnderlayParams) {
+func validateUnderlayInNS(g Gomega, ns netns.NsHandle, skipMainNicCheck bool, params UnderlayParams) {
```
let's call this CheckUnderlayNic

fedepaol commented on internal/hostnetwork/underlay_test.go:
```diff
@@ -72,7 +72,7 @@ var _ = Describe("Underlay configuration should work when", func() {
 
 		Eventually(func(g Gomega) {
-			validateUnderlayInNS(g, testNs, params)
+			validateUnderlayInNS(g, testNs, false, params)
```
let's use a const instead of false. Something like "CheckUnderlay"

fedepaol commented on cmd/hostcontroller/main.go:
```diff
@@ -83,6 +84,7 @@ func main() {
 		"the location of the frr configuration file")
 	flag.IntVar(&args.reloadPort, "reloadport", 9080, "the port of the reloader process")
 	flag.StringVar(&args.criSocket, "crisocket", "/containerd.sock", "the location of the cri socket")
+	flag.BoolVar(&args.multusEnabled, "multus-enabled", false, "Whether Multus is enabled")
```
same for all the variables / parameters along the same line
